File: java/206.java
public class ListNode{
    int val;
    ListNode next;
    ListNode(int x){
        val = x
    }
}

class Solution{
	// 迭代
    public ListNode reverseList(ListNode head){
        ListNode pre = null;
	ListNode cur = head;
	ListNode temp;
	while(cur != null){
	    temp = cur.next;
	    cur.next = pre;
	    pre = cur;
	    cur = temp;
	}
	return pre;
    }

    //迭代
//	public ListNode reverseList(ListNode head) {
//		if (head == null || head.next == null)
//			return head;
//		ListNode new_head = reverseList(head.next);
//		head.next.next = head;
//		head.next = null;
//		return new_head;
//	}
}

File: python/206.py
# class ListNode:
#     __init__(self, x):
#         self.val = x
#         self.next = None
#

class Solution:

    # 迭代
    # def reverseList(self, head: ListNode) -> ListNode:
    #     pre, cur = None, head
    #     while(cur):
    #         cur.next, pre, cur = pre, cur, cur.next
    #     return pre
    # 递归
    def reverseList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        new_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return new_head
